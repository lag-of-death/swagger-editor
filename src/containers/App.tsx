import { IRuleResult } from "@stoplight/spectral";
import * as React from "react";
import { connect } from "react-redux";
import { updateText } from "../actions";
import { TreeView } from "../components/TreeView";
import { ISpecPart } from "../components/TreeView/interfaces";
import { Store } from "../shared";
import { diagnose } from "../util/linting";
import IssuesViewer from "./IssuesViewer";

type Event = { target: { value: string } };

import styled, { css } from "styled-components";
import { border, borderRadius, boxShadow, padding } from "./shared";

const shared = css`
  ${border}
  ${borderRadius}
  ${boxShadow}
`;

const JSONValidity = styled.div`
  ${shared}
  ${padding}
  margin: 10px;
  text-align: center;
`;

const TextEditor = styled.textarea`
  ${padding}
  display: flex;
  flex-grow: 20;
  margin: 10px;
  ${shared}
`;

const Editor = styled.div`
  display: flex;
  flex-flow: column;
  flex-grow: 10;
  flex-shrink: 1;
  flex-basis: 0;
`;

const Container = styled.div`
  display: flex;
  flex-grow: 1;
  height: 100%;
`;

const TreeViewContainer = styled.div`
  overflow: auto;
  flex-grow: 5;
  flex-shrink: 1;
  flex-basis: 0;
  margin: 10px;
  ${shared}
`;

export const App = (
  {spec, issues, text, onChange, isJSONValid}:
    { onChange: (evt: Event) => null, text: string, issues: IRuleResult[], spec: ISpecPart, isJSONValid: boolean },
) => {
  const stringifiedJSON = JSON.stringify(spec, null, 2);

  return (
    <Container>
      <TreeViewContainer>
        <TreeView
          spec={spec}
          diagnostics={diagnose(spec)}/>
      </TreeViewContainer>

      <Editor>
        <IssuesViewer/>
        <JSONValidity>
          {isJSONValid ? "JSON VALID" : "JSON NOT VALID"}
        </JSONValidity>
        <TextEditor
          value={text ? text : (Object.keys(spec).length ? stringifiedJSON : "")}
          onChange={onChange}/>
      </Editor>
    </Container>
  );
};

export default connect((state: Store) => {
  return {
    isJSONValid: state.isJSONValid,
    issues: state.issues,
    spec: state.spec,
    text: state.text,
  };
}, (dispatch) => {
  return {
    onChange: ({target: {value}}: Event) => {
      dispatch(updateText(value));
    },
  };
})(App);
