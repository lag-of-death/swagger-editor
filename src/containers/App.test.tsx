import { shallow } from "enzyme";
import * as React from "react";
import { App } from "./App";

describe("App", () => {
  test("rendering", () => {
    const component = shallow(
      <App text="some text" isJSONValid={true} onChange={() => null} issues={[]} spec={({})}/>,
    );

    expect(component).toMatchSnapshot();
  });
});
