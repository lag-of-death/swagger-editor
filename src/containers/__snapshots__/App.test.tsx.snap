// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering 1`] = `
<styled.div>
  <styled.div>
    <TreeView
      diagnostics={
        Array [
          Object {
            "message": "info.contact is not truthy",
            "name": "info-contact",
            "path": Array [
              "info",
              "contact",
            ],
            "severity": 40,
            "severityLabel": "warn",
            "summary": "Info object should contain \`contact\` object.",
          },
          Object {
            "message": "info.description is not truthy",
            "name": "info-description",
            "path": Array [
              "info",
              "description",
            ],
            "severity": 40,
            "severityLabel": "warn",
            "summary": "OpenAPI object info \`description\` must be present and non-empty string.",
          },
          Object {
            "message": "should have required property 'swagger'",
            "name": "oas2-schema",
            "path": Array [],
            "severity": 50,
            "severityLabel": "error",
            "summary": "Validate structure of OpenAPIv2 specification.",
          },
          Object {
            "message": "host is not truthy",
            "name": "api-host",
            "path": Array [
              "host",
            ],
            "severity": 40,
            "severityLabel": "warn",
            "summary": "OpenAPI \`host\` must be present and non-empty string.",
          },
          Object {
            "message": "schemes does not exist",
            "name": "api-schemes",
            "path": Array [
              "schemes",
            ],
            "severity": 40,
            "severityLabel": "warn",
            "summary": "OpenAPI host \`schemes\` must be present and non-empty array.",
          },
        ]
      }
      spec={Object {}}
    />
  </styled.div>
  <styled.div>
    <Connect(Component) />
    <styled.div>
      JSON VALID
    </styled.div>
    <styled.textarea
      onChange={[Function]}
      value="some text"
    />
  </styled.div>
</styled.div>
`;
